proto/health.proto

```diff
@@ -0,0 +1,563 @@
+  message Faults {
+    // A list of faults or errors currently present on the device.
+    // The presence of any current faults will result in an ABNORMAL state.
+    // If the device is healthy this list should be empty.
+    repeated Error current_faults = 1;
+
+    // todo: maybe some way to list the possible faults the device can report?
+  }
+
+  // Details about the check being performed.
+  // Optional, but strongly recommended.
+  // HealthChecks should not change their type of check after creation.
+  oneof check {
+    // This check is comparing a value against known good/bad values described by this field.
+    Bounds bounds = 15;
+    // This check is looking for faults or alarms on the device.
+    Faults faults = 16;
+  }
+
+  // An acknowledgement for a change in health state.
+  // Some changes in health state require acknowledgement from another person or system before they can be considered
+  // resolved.
+  //
+  // For example, a fire detection system may detect smoke, which will cause the health check to change to ABNORMAL.
+  // Even if the smoke clears and the check returns to NORMAL, the system should still require a person to investigate
+  // the abnormality.
+  // Requiring an acknowledgement allows the system to track whether the abnormality has been investigated,
+  // or whether the system is still in a state that requires attention.
+  //
+  // Each check can specify whether it expects an acknowledgement independently for each health state transition.
+  // The possible transitions are represented by the HealthChange enum.
+  // Changing the expected acks will not change any existing acks, but will change the required acks for future changes
+  // in health state.
+  //
+  // The Ack itself can come from user interaction with a Smart Core app, via this API,
+  // from an interaction with a system managing a device, like a headend workstation,
+  // or from physical interaction with the device, like pressing a reset button.
+  // The source field should identify the source of the acknowledgement.
+  message Ack {
+    google.protobuf.Timestamp ack_time = 1;
+    // The actor that has acknowledged the change in health, if known.
+    smartcore.bos.Actor actor = 2;
+    // The source of the acknowledgement.
+    // This should identify the app or system that acknowledged the change in health.
+    // For example, "smartcore.app.ops", "reset button", or "device".
+    string source = 3;
+  }
+  // Health state transitions as different bit mask values.
+  // See Ack for details.
+  enum HealthChange {
+    HEALTH_CHANGE_UNSPECIFIED = 0;
+    // A change from an unhealthy state to the healthy state.
+    TO_HEALTHY = 1;
+    // A change to a reliability state that is not RELIABLE.
+    TO_UNRELIABLE = 2;
+    // A change to a check state that is not NORMAL.
+    TO_ABNORMAL = 4;
+  }
+  // A bit mask of health state transitions that will require acknowledgement.
+  // See Ack for details.
+  int32 ack_expected = 20;
```

Comment:
Bit masks don't strike me as idiomatic in protobuf. Is there any precedent you referenced when designing this? Is there a reason a repeated enum would be inadequate?

Reply:
I'm hoping to punt the Ack stuff until after all this lands. I don't have any examples, there's no API to interact with it, and questions like this need more attention that in the whole PR. I wasn't sure how to set up the PR without this bit without adding Git overhead.